File: usaco/section_5.2/snail.py
"""
ID: huangch7
LANG: PYTHON3
TASK: snail
"""
import math

task_name = "snail"
INF = 100000000


def get_list_from_line(dtype=int):
    return [dtype(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()



fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

N, B = get_list_from_line()
barriers = set()
for _ in range(B):
    line = fin.readline()
    col = ord(line[0]) - ord('A')
    row = int(line[1:]) - 1
    barriers.add((row, col))

visited = [[0] * N for _ in range(N)]
visited[0][0] = 1
best = 0

direction_map = {
    0: [0, 1], # >
    1: [1, 0], # V
    2: [0, -1],# <
    3: [-1, 0] # ^
}
# cache = {}

def display_visited():
    for i in range(len(visited)):
        print("".join([str(x) for x in visited[i]]))
    print('-' * 10)

total = 0
def move(row, col, direction):
    global total
#    print(row, col, direction)
#    display_visited()
    
#    total += 1
#    if total > 10:
#        return -1
#     if (row, col, direction) in cache:
#        return cache[(row, col, direction)]
    cur = [row, col]
    cur_steps = 0
    is_end = False
    buff = []
    while True:
        cur[0] += direction_map[direction][0]
        cur[1] += direction_map[direction][1]
        if (cur[0], cur[1]) in barriers:
            break
        if (cur[0] < 0 or cur[0] >= N):
            break
        if (cur[1] < 0 or cur[1] >= N):
            break
        if visited[cur[0]][cur[1]]:
            is_end = True
            break
        visited[cur[0]][cur[1]] = 1
        cur_steps += 1
        buff.append(tuple(cur))
    cur[0] -= direction_map[direction][0]
    cur[1] -= direction_map[direction][1]
    if not is_end and (cur[0] != row or cur[1] != col):
        cur_steps += max(move(cur[0], cur[1], (direction + 1) % 4),
                         move(cur[0], cur[1], (direction - 1) % 4))
    # backforward the visited
#     print(buff)
    for r, c in buff:
        visited[r][c] = 0

#    cache[(row, col, direction)] = cur_steps
    return cur_steps

best = move(0, 0, 0)
# display_visited()
best = max(move(0, 0, 1), best)

fout.write(f"{best + 1}\n")


fout.close()

File: usaco/training/section_5.2/snail.cpp
/*
ID: huangch7
LANG: C++
TASK: snail
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

const int INF = 100000000;

int N, B;
set<pair<int, int> > barriers;
vector<vector<int> > visited;

int direction_map[][2] = {
    {0, 1},   // >
    {1, 0},   // V
    {0, -1},  // <
    {-1, 0}   // ^
};

void displayVisited() {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cout << visited[i][j];
        }
        cout << endl;
    }
    cout << "----------" << endl;
}

int move(int row, int col, int direction) {
    int total = 0;
    vector<pair<int, int>> buffer;
    int cur_steps = 0;
    bool is_end = false;
    vector<int> cur = {row, col};

    while (true) {
        cur[0] += direction_map[direction][0];
        cur[1] += direction_map[direction][1];

        if (barriers.count(make_pair(cur[0], cur[1])) || cur[0] < 0 || cur[0] >= N || cur[1] < 0 || cur[1] >= N) {
            break;
        }

        if (visited[cur[0]][cur[1]]) {
            is_end = true;
            break;
        }

        visited[cur[0]][cur[1]] = 1;
        cur_steps++;
        buffer.emplace_back(make_pair(cur[0], cur[1]));
    }

    cur[0] -= direction_map[direction][0];
    cur[1] -= direction_map[direction][1];

    if (!is_end && (cur[0] != row || cur[1] != col)) {
        cur_steps += max(move(cur[0], cur[1], (direction + 1) % 4), move(cur[0], cur[1], (direction + 3) % 4));
    }

    for (const auto& pos : buffer) {
        visited[pos.first][pos.second] = 0;
    }

    return cur_steps;
}

int main() {
    ifstream fin("snail.in");
    ofstream fout("snail.out");

    fin >> N >> B;
    visited.resize(N, vector<int>(N, 0));

    for (int i = 0; i < B; ++i) {
        char c;
        int row;
        fin >> c >> row;
        int col = c - 'A';
        barriers.emplace(make_pair(row - 1, col));
    }

    visited[0][0] = 1;
    int best = move(0, 0, 0);
    best = max(move(0, 0, 1), best);

    fout << best + 1 << endl;

    fin.close();
    fout.close();

    return 0;
}
